# 실습 5: 센서 590개 진단표 만들기
- 상황: 관리도로 한 개는 봤는데 나머지 589개가 남았다
- 목표: 열 하나를 한 줄로 요약해 쓸 수 없는 열을 걸러낸다

## Step 0. 폴더와 노트북 만들고 데이터 불러오기

In [4]:
import pandas as pd

df = pd.read_csv("../../data/04_secom.csv")

print("행 수, 열 수:", df.shape)


행 수, 열 수: (1567, 592)


## Step 1. 센서 열 진단표 만들기
센서 열마다 빈칸 비율·고유값 개수·표준편차·최솟값·최댓값을 계산해 `sensor_diagnosis` 표로 만들고, 빈칸 비율이 높은 순으로 정렬한다.

In [5]:
sensor_cols = [c for c in df.columns if c.startswith("sensor_")]

행목록 = []
for c in sensor_cols:
    빈칸비율 = df[c].isna().mean() * 100
    고유값개수 = df[c].nunique()
    표준편차 = df[c].std()
    최솟값 = df[c].min()
    최댓값 = df[c].max()
    행목록.append({
        "열 이름": c,
        "빈칸 비율(%)": round(빈칸비율, 2),
        "고유값 개수": 고유값개수,
        "표준편차": 표준편차,
        "최솟값": 최솟값,
        "최댓값": 최댓값,
    })

sensor_diagnosis = pd.DataFrame(행목록)
sensor_diagnosis = sensor_diagnosis.sort_values("빈칸 비율(%)", ascending=False).reset_index(drop=True)

print("표 전체 줄 수:", len(sensor_diagnosis))
sensor_diagnosis.head(10)


표 전체 줄 수: 590


,열 이름,빈칸 비율(%),고유값 개수,표준편차,최솟값,최댓값
0,sensor_293,91.19,92,0.011494,0.0041,0.0831
1,sensor_294,91.19,138,137.692483,82.3233,879.2260
2,sensor_159,91.19,138,406.848810,234.0996,2505.2998
3,sensor_158,91.19,128,0.039538,0.0118,0.2876
4,sensor_493,85.58,226,1.759262,4.8882,21.0443
5,sensor_086,85.58,97,0.002928,0.1053,0.1184
6,sensor_359,85.58,20,0.000395,0.0017,0.0047
7,sensor_221,85.58,69,0.001989,0.0057,0.0240
8,sensor_245,64.96,66,0.084618,0.0003,1.9844
9,sensor_518,64.96,543,4.890663,0.2880,113.2758


### 결과 정리 (실행 결과 기준)

표 전체 줄 수: **590줄** (센서 열 개수만큼)

빈칸 비율 상위 10개:

| 열 이름 | 빈칸 비율(%) | 고유값 개수 | 표준편차 | 최솟값 | 최댓값 |
|---|---|---|---|---|---|
| sensor_293 | 91.19 | 92 | 0.011494 | 0.0041 | 0.0831 |
| sensor_294 | 91.19 | 138 | 137.692483 | 82.3233 | 879.2260 |
| sensor_159 | 91.19 | 138 | 406.848810 | 234.0996 | 2505.2998 |
| sensor_158 | 91.19 | 128 | 0.039538 | 0.0118 | 0.2876 |
| sensor_493 | 85.58 | 226 | 1.759262 | 4.8882 | 21.0443 |
| sensor_086 | 85.58 | 97 | 0.002928 | 0.1053 | 0.1184 |
| sensor_359 | 85.58 | 20 | 0.000395 | 0.0017 | 0.0047 |
| sensor_221 | 85.58 | 69 | 0.001989 | 0.0057 | 0.0240 |
| sensor_245 | 64.96 | 66 | 0.084618 | 0.0003 | 1.9844 |
| sensor_518 | 64.96 | 543 | 4.890663 | 0.2880 | 113.2758 |

빈칸 비율이 가장 높은 4개 열(sensor_293, 294, 159, 158)은 91.19%가 비어 있어, 값이 채워진 행이 전체 1,567행 중 138행뿐이다.

## Step 2. 빈칸 비율 낮은 순 정렬 + 고유값 1개·표준편차 0인 열 세기

In [6]:
# 빈칸 비율 낮은 순으로 정렬
낮은순 = sensor_diagnosis.sort_values("빈칸 비율(%)", ascending=True).reset_index(drop=True)
print("빈칸 비율 낮은 순 10줄")
display(낮은순.head(10))

# 서로 다른 값이 1개뿐인 열 개수
고유값1개 = (sensor_diagnosis["고유값 개수"] == 1).sum()
print()
print("서로 다른 값의 개수가 1개인 열:", 고유값1개, "개")

# 표준편차가 0인 열 개수
표준편차0 = (sensor_diagnosis["표준편차"] == 0).sum()
print("표준편차가 0인 열:", 표준편차0, "개")


빈칸 비율 낮은 순 10줄


,열 이름,빈칸 비율(%),고유값 개수,표준편차,최솟값,최댓값
0,sensor_121,0.0,1269,0.124304,5.1259,7.5220
1,sensor_521,0.0,1536,5.702366,0.3121,111.7365
2,sensor_522,0.0,9,103.122996,0.0000,1000.0000
3,sensor_523,0.0,1562,7.104435,2.6811,137.9838
4,sensor_524,0.0,1040,4.147581,0.0258,111.3330
5,sensor_525,0.0,1543,20.663414,1.3104,818.0005
6,sensor_527,0.0,1514,0.958428,0.1705,8.2037
7,sensor_572,0.0,811,0.275112,0.9802,2.7395
8,sensor_394,0.0,953,0.038408,0.0342,0.2994
9,sensor_021,0.0,552,0.016737,1.1797,1.4534



서로 다른 값의 개수가 1개인 열: 116 개
표준편차가 0인 열: 116 개


### 결과 정리 (실행 결과 기준)

빈칸 비율 낮은 순(빈칸 0%인 열들) 10개:

| 열 이름 | 빈칸 비율(%) | 고유값 개수 | 표준편차 | 최솟값 | 최댓값 |
|---|---|---|---|---|---|
| sensor_523 | 0.0 | 1562 | 7.104435 | 2.6811 | 137.9838 |
| sensor_528 | 0.0 | 1549 | 1.888698 | 2.1700 | 14.4479 |
| sensor_527 | 0.0 | 1514 | 0.958428 | 0.1705 | 8.2037 |
| sensor_525 | 0.0 | 1543 | 20.663414 | 1.3104 | 818.0005 |
| sensor_521 | 0.0 | 1536 | 5.702366 | 0.3121 | 111.7365 |
| sensor_524 | 0.0 | 1040 | 4.147581 | 0.0258 | 111.3330 |
| sensor_522 | 0.0 | 9 | 103.122996 | 0.0000 | 1000.0000 |
| sensor_494 | 0.0 | 572 | 0.973948 | 0.8330 | 9.4024 |
| sensor_393 | 0.0 | 109 | 0.002956 | 0.0005 | 0.0229 |
| sensor_496 | 0.0 | 964 | 3.260019 | 1.7720 | 107.6926 |

- **서로 다른 값의 개수가 1개인 열: 116개**
- **표준편차가 0인 열: 116개**

두 조건의 개수가 똑같이 116개로 나왔다 — 값이 하나뿐이면 흩어질 여지가 없어 표준편차도 자연히 0이 되므로, 사실상 같은 열들을 가리키는 결과로 보인다.

[내가 정한 1차 선별 기준]<br>
1. 빈칸 비율이 [50]% 이상인 열은 버린다<br>
→ 왜 이 선인가 : [절반 넘게 비어 있으면 남은 값으로 채워 넣어도 지어낸 값에 가깝다]<br>
2. 값 종류가 1개인 열은 버린다 (상수열)<br>
→ 왜 : [처음부터 끝까지 같은 값이면 양품과 불량을 가르는 데 아무 기여를 못 한다]<br>
3. 표준편차가 [0.001] 이하인 열은 버린다<br>
→ 왜 이 선인가 : [정렬해서 보니 이 언저리에서 값이 확 벌어졌다. 그 아래는 사실상 안 변하는 열로 봤다]<br>
<br>
+ 내 공정 감각으로 한 줄 : [설비를 세우면 그 구간 값이 통째로 빈다. 그래서 빈칸이 많다고<br>
무조건 못 쓸 열은 아닐 수 있어 50%를 낮추지 않았다]